File: example_clients/python/tokenbucket_client.py
import socket
import time

class TokenBucketClient:
    def __init__(self, hostname, port):
        self.hostname = hostname
        self.port = port

    def consume(self, bucket_name, callback):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((self.hostname, self.port))

            while True:
                s.sendall(f"CONSUME {bucket_name}\n".encode())
                response = s.recv(1024).decode().strip()
                status, *rest = response.split(' ')

                if status == 'OK':
                    return callback()
                elif status == 'WAIT':
                    sleep_time = float(rest[0])
                    time.sleep(sleep_time)
                else:
                    raise Exception(f"Unknown response from server: {response}")

# Example usage:
# client = TokenBucketClient('localhost', 2000)
# client.consume('foo', lambda: print("Running some operation..."))

File: example_clients/ruby/tokenbucket_client.rb
require 'socket'

# an example client of this service
class TokenBucketClient
  def initialize(hostname:, port:)
    @server = TCPSocket.new(hostname, port)
  end

  def consume(bucket_name)
    loop do
      @server.puts("CONSUME #{bucket_name}")
      response = @server.gets.chomp
      status, *rest = response.split

      case status
      when 'OK'
        return yield
      when 'WAIT'
        sleep_time, *_ = rest
        sleep(sleep_time.to_f)
      else
        raise "Unknown response from server: #{response}"
      end
    end
  ensure
    @server.close
  end
end

# Example usage:
# client = TokenBucketClient.new(hostname: 'localhost', port: 2000)
# client.consume('foo') do
#  # do something
# end